Src/HALALMock/Services/Communication/FDCAN/FDCAN.cpp

```diff
@@ -89,43 +92,52 @@ bool FDCAN::transmit(uint8_t id, uint32_t message_id, const char* data, FDCAN::D
 		return false;
 	}
 
-	instance->tx_header.Identifier = message_id;
+    size_t buffer_len = sizeof(message_id)+ sizeof(dlc) + dlc;
+    char* temp_data = new char[buffer_len];
 
-	if (dlc != FDCAN::DLC::DEFAULT) {
-		instance->tx_header.DataLength = dlc;
-	}
+    temp_data[0] = (message_id >> 24);
+    temp_data[1] = (message_id >> 16);
+    temp_data[2] = (message_id >> 8);
+    temp_data[3] = message_id;
 
-	HAL_StatusTypeDef error = HAL_FDCAN_AddMessageToTxFifoQ(instance->hfdcan, &instance->tx_header, (uint8_t*)data);
+	temp_data[4] = ((uint32_t)dlc >> 24);
+	temp_data[5] = ((uint32_t)dlc >> 16);
+	temp_data[6] = ((uint32_t)dlc >> 8);
+	temp_data[7] = (uint32_t)dlc;
 
-	if (error != HAL_OK) {
+	memcpy((temp_data + sizeof(message_id)+ sizeof(dlc)), data, dlc);
+		
+	if(send(instance->socket,temp_data, strlen(data))<0){
 		ErrorHandler("Error sending message with id: 0x%x by FDCAN %d", message_id, instance->fdcan_number);
+		delete[] temp_data;
 		return false;
 	}
 
+	delete[] temp_data;
 	return true;
 }
 
-void HAL_FDCAN_RxFifo0Callback(FDCAN_HandleTypeDef *hfdcan, uint32_t RxFifo0ITs){
-	FDCAN::read(FDCAN::handle_to_id[hfdcan],&packet);
-	if(packet.identifier == FDCAN::ID::FAULT_ID){
-		ErrorHandler("FAULT PROPAGATED via CAN");
-	}
-}
 
 bool FDCAN::read(uint8_t id, FDCAN::Packet* data){
-	if (not FDCAN::registered_fdcan.contains(id)) {
+if (not FDCAN::registered_fdcan.contains(id)) {
 		ErrorHandler("There is no FDCAN registered with id: %d.", id);
 		return false;
 	}
+	FDCAN::Instance* instance = registered_fdcan[id];
 
 	if(!FDCAN::received_test(id)) {
 		return false;
 	}
-	FDCAN_RxHeaderTypeDef header_buffer = FDCAN_RxHeaderTypeDef();
-	HAL_FDCAN_GetRxMessage(FDCAN::registered_fdcan.at(id)->hfdcan, FDCAN::registered_fdcan.at(id)->rx_location, &header_buffer, data->rx_data.data());
+	
+	if(recv(instance->socket, data->rx_data, 64,0)<0){
+		ErrorHandler("Error receiving message by FDCAN %d", instance->fdcan_number);
+		return false;
+	}
 
-	data->identifier = header_buffer.Identifier;
-	data->data_length = static_cast<FDCAN::DLC>(header_buffer.DataLength);
+	data->identifier = (data->rx_data[0] | data->rx_data[1] | data->rx_data[2] | data->rx_data[3]);
+	data->data_length = static_cast<FDCAN::DLC>((data->rx_data[4] | data->rx_data[5]| data->rx_data[6] | data->rx_data[7]));
+	std::copy(data->rx_data.begin() + 8, data->rx_data.end(), data->rx_data.begin());
```

Comment:
I think this is undefined behaviour, 


Check cppreference: [https://en.cppreference.com/w/cpp/algorithm/copy](https://en.cppreference.com/w/cpp/algorithm/copy)
I quote:

> Copies all elements in the range [first, last) starting from first and proceeding to last.
>  If d_first is in [first, last), the behavior is undefined. In this case, [std::copy_backward](https://en.cppreference.com/w/cpp/algorithm/copy_backward) may be used instead.

I think the easiest thing is to create an auxiliary reception buffer and just copy from there, unparsing the id, the DLC and the data. Besides from this, good job!